src/Analyser/MutatingScope.php

```diff
@@ -4858,7 +4858,15 @@ private static function generalizeType(Type $a, Type $b): Type
 			$resultTypes[] = TypeCombinator::union(...$integerRanges['b']);
 		}
 
-		return TypeCombinator::union(...$resultTypes, ...$otherTypes);
+		$accessoryTypes = array_map(
+			static fn (Type $type): Type => $type->generalize(GeneralizePrecision::moreSpecific()),
+			TypeUtils::getAccessoryTypes($a),
```

Comment:
What about the accessory types that are only in b?

Reply:
those lead to regressions by either falsely narrowing or widening types in loops and closures. e.g. as described in https://github.com/phpstan/phpstan-src/pull/1732#issuecomment-1255903884

Reply:
and using only a or types in a AND b doesn't change anything, so I just kept it simple and went with a types.

Reply:
Alright, interesting!

Reply:
yeah it confused me quite a bit, but since everything was OK and the newly added tests are not breaking either and the generalisation order for loops and closures is important I just went with it at some point tbh